pkg/reconciler/common/prune.go

```diff
@@ -21,75 +21,199 @@ import (
-		logger.Error("failed to create cronjob ", err)
 
+	if pruningNamespaces.commonScheduleNs != nil && len(pruningNamespaces.commonScheduleNs) > 0 {
+		jobs := pruner.createAllJobs(pruningNamespaces.commonScheduleNs)
+
+		if err := pruner.createCronJob(tC.Spec.TargetNamespace, tC.Spec.Pruner.Schedule, jobs, ownerRef); err != nil {
+			logger.Error("failed to create cronjob ", err)
+		}
+	}
+	if pruningNamespaces.uniqueScheduleNS != nil {
+		for ns, con := range pruningNamespaces.uniqueScheduleNS {
+			jobs := pruner.createJobs(con, ns)
+
+			if err := pruner.createCronJob(ns, con.config.Schedule, jobs, ownerRef); err != nil {
+				logger.Errorf("failed to create cronjob in ns %s:", ns, err)
+			}
+		}
 	}
 
 	return nil
 }
 
-func GetPrunableNamespaces(k kubernetes.Interface, ctx context.Context) ([]string, error) {
+func prunableNamespaces(k kubernetes.Interface, ctx context.Context, defaultPruneConfig v1alpha1.Prune) (pruningNs, error) {
 	nsList, err := k.CoreV1().Namespaces().List(ctx, v1.ListOptions{})
 	if err != nil {
-		return nil, err
+		return pruningNs{}, err
 	}
-
-	var allNameSpaces []string
+	var prunableNs pruningNs
+	commonSchedule := make(map[string]*pruneConfigPerNS)
+	uniqueSchedule := make(map[string]*pruneConfigPerNS)
 	re := regexp.MustCompile(NamespaceIgnorePattern)
 	for _, ns := range nsList.Items {
 		if ignore := re.MatchString(ns.GetName()); ignore {
 			continue
 		}
-		allNameSpaces = append(allNameSpaces, ns.Name)
+		nsAnnotations := ns.GetAnnotations()
+		//skip all the namespaces if annotated with prune skip
+		if nsAnnotations[pruneSkip] != "" && nsAnnotations[pruneSkip] == "true" {
+			continue
+		}
+		pc := &pruneConfigPerNS{
+			config:               v1alpha1.Prune{},
+			relatedPipelineNames: []string{},
+			relatedTaskNames:     []string{},
+		}
+
+		if nsAnnotations[pruneResourcesPipelineNames] != "" {
+			pc.relatedPipelineNames = strings.Split(nsAnnotations[pruneResourcesPipelineNames], ",")
+		}
+		if nsAnnotations[pruneResourcesTaskNames] != "" {
+			pc.relatedTaskNames = strings.Split(nsAnnotations[pruneResourcesTaskNames], ",")
+		}
+		if nsAnnotations[pruneResources] != "" {
+			pc.config.Resources = strings.Split(nsAnnotations[pruneResources], ",")
+		} else {
+			pc.config.Resources = defaultPruneConfig.Resources
+		}
+		if nsAnnotations[pruneKeepSince] != "" {
+			keepsince, _ := strconv.Atoi(nsAnnotations[pruneKeepSince])
+			uintKeepSince := uint(keepsince)
+			pc.config.KeepSince = &uintKeepSince
+			pc.config.Keep = nil
+		} else if defaultPruneConfig.KeepSince != nil && nsAnnotations[pruneKeep] == "" {
+			pc.config.KeepSince = defaultPruneConfig.KeepSince
+		}
+		if nsAnnotations[pruneKeep] != "" {
+			keep, _ := strconv.Atoi(nsAnnotations[pruneKeep])
+			uintKeep := uint(keep)
+			pc.config.Keep = &uintKeep
+			pc.config.KeepSince = nil
+		} else if defaultPruneConfig.Keep != nil && nsAnnotations[pruneKeepSince] == "" {
+			pc.config.Keep = defaultPruneConfig.Keep
+		}
```

Comment:
we have to add `pruneStrategy` logic here
as mentioned in  https://github.com/tektoncd/operator/pull/410/files#r712087264